src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -0,0 +1,350 @@
+    Object valueContent = isKey ? record.value() : newSFContent;
+    return new SinkRecord(
+        record.topic(),
+        record.kafkaPartition(),
+        keySchema,
+        keyContent,
+        valueSchema,
+        valueContent,
+        record.kafkaOffset(),
+        record.timestamp(),
+        record.timestampType(),
+        record.headers());
+  }
+
+  public void insertBufferedRows() {
+    PartitionBuffer intermediateBuffer = null;
+    bufferLock.lock();
+    try {
+      intermediateBuffer = streamingBuffer;
+      streamingBuffer = new StreamingBuffer();
+    } finally {
+      // release lock
+      bufferLock.unlock();
+    }
+    // intermediate buffer can be empty here if time interval reached but kafka produced no records.
+    if (intermediateBuffer.isEmpty()) {
+      LOGGER.info("No Rows Buffered, returning;");
+      this.previousFlushTimeStampMs = System.currentTimeMillis();
+      return;
+    }
+    InsertValidationResponse response =
+        channel.insertRows(
+            (Iterable<Map<String, Object>>) intermediateBuffer.getData(),
+            Long.toString(intermediateBuffer.getLastOffset()));
+
+    // need to update the flush time(last I called insertrows)
+    this.previousFlushTimeStampMs = System.currentTimeMillis();
+
+    // should we retry if response.hasErrors() == true?
+    // how can we retry if buffer was cleared?
+    // lets have an exponential back off and jitter retry policy
+
+    if (response.hasErrors()) {
+      LOGGER.error(
+          "Inserted rows API response has errors for rows with offset start:{} and end:{}",
+          intermediateBuffer.getFirstOffset(),
+          intermediateBuffer.getLastOffset());
+      response
+          .getInsertErrors()
+          .forEach(
+              insertError -> {
+                LOGGER.error(
+                    "Insert row Error message:{}, with ex:{}",
+                    insertError.getMessage(),
+                    insertError.getException().getMessage());
+              });
+    }
+  }
+
+  // should we rely on precommits frequency to give us the last committed offset token?
+  // or should we have a BG thread polling on every channel to update the Atomic long and simply
+  // return the atomic long's current value
+  // For now we will rely on preCommit API to fetch the inserted offsets into Snowflake
+  // TODO: SNOW-529755 POLL committed offsets in backgraound thread
+  public long getCommittedOffset() {
+    LOGGER.info(
+        "Fetching last committed offset for partition channel:{}",
+        this.channel.getFullyQualifiedName());
+    String lastOffsetCommitted = channel.getLatestCommittedOffsetToken();
+    LOGGER.info(
+        "Last committed offset for partition channel:{} is :{}",
+        this.channel.getFullyQualifiedName(),
+        lastOffsetCommitted);
+    if (Strings.isNullOrEmpty(lastOffsetCommitted)
+        || lastOffsetCommitted.equalsIgnoreCase("null")) {
+      return committedOffset.get();
```

Comment:
could we have a null or empty offset for kafka?

Reply:
yes.. for first time. default is 0 to begin with when we return in precommit..

Reply:
precommit = API kafka calls us immediately after put. We send offset no safe to commit back to kafka. 

This committedOffset = offset Ingested in SF + 1

Idea behind sending +1 is that kafka will send us offset starting from that number because we already ingested previous one.